### PR TITLE
C27 refactor and integrate front backend

### PR DESCRIPTION
**Pull request by Alexkszzz**
## Change Summary
 * Frontend is redesigned using data-binded render
 * Backend is redesigned 
 * changing variables name for consistency
 * Added Frontend functions to integrate with Backend
 * Added data loading

### CODE REVIEW

File: app/app/static/custom/ourscript.js [hunk 1]
```diff
@@ -4,85 +4,215 @@
...
+	try{
+		const response = await fetch("add_entry", {
+			method: "POST",
+			body: JSON.stringify(data),
+			headers: {
+				'Content-Type': 'application/json'
+			},
+		})
+		const dataResponse = await response.json()
+		console.log(dataResponse)
```
Comment (frinzekt) [inline app/app/static/custom/ourscript.js:30]: Remove console log

File: app/app/static/custom/ourscript.js [hunk 2]
```diff
@@ -4,85 +4,215 @@
...
+		// There's an error
+		console.log(error)
+	}
+}
+
+const moveToSessionList = async(data) => {
+	// This function move the data (object - of the student details) to the in Session table
+	// And rerenders the table
+	try{
+		const response = await fetch("/update_entry/" + data.id, {
```
Comment (frinzekt) [inline app/app/static/custom/ourscript.js:47]: Use JavaScript string template

File: app/app/static/custom/ourscript.js [hunk 3]
```diff
@@ -4,85 +4,215 @@
...
+	})
+
+	hideAddToQueueForm();
 });
 
-function deleteRow(x) {
-	$(x).parents('tr').remove();
+function deleteRow(id, status) {
+	const closureFunction = (currentElement) => {
+		console.log(currentElement)
```
Comment (frinzekt) [inline app/app/static/custom/ourscript.js:175]: Remove console log

File: app/app/static/custom/ourscript.js [hunk 4]
```diff
@@ -4,85 +4,215 @@
...
+		queue,
+		enquiry
+	})
+
+	hideAddToQueueForm();
 });
 
-function deleteRow(x) {
-	$(x).parents('tr').remove();
+function deleteRow(id, status) {
```
Comment (frinzekt) [inline app/app/static/custom/ourscript.js:173]: Give a brief description of what this function does

Comment (frinzekt) [reply app/app/static/custom/ourscript.js:173]: You also have to say that this is a closure function

File: app/app/static/custom/ourscript.js [hunk 5]
```diff
@@ -4,85 +4,215 @@
...
+			id,
+			status
+		})
+}
+return closureFunction
 }
 
-function addSessionToTeam(team, id){
+function finishRow(id, status) {
+	const closureFunction = (currentElement) => {
```
Comment (frinzekt) [inline app/app/static/custom/ourscript.js:186]: Give brief description of what this function does

File: app/app/static/custom/ourscript.js [hunk 6]
```diff
@@ -4,85 +4,215 @@
...
+		terminateSession({
+			id,
+			status
+		})
+}
+return closureFunction
 }
 
-function addSessionToTeam(team, id){
+function finishRow(id, status) {
```
Comment (frinzekt) [inline app/app/static/custom/ourscript.js:185]: Both finishRow and Delete row has the same content in function except for status. Consider just using the parameter at the top to simplify the two?

File: app/app/static/custom/ourscript.js [hunk 7]
```diff
@@ -4,85 +4,215 @@
...
+	const closureFunction = (currentElement) => {
+		console.log(currentElement)	
+		status = "Completed"
```
Comment (frinzekt) [inline app/app/static/custom/ourscript.js:188]: What's the difference between this variable status and the function parameter status?

File: app/app/static/custom/ourscript.js [hunk 8]
```diff
@@ -4,85 +4,215 @@
...
+			status
+		})
+}
+return closureFunction
 }
 
-function addSessionToTeam(team, id){
+function finishRow(id, status) {
+	const closureFunction = (currentElement) => {
+		console.log(currentElement)
```
Comment (frinzekt) [inline app/app/static/custom/ourscript.js:187]: Remove console log

File: app/app/static/custom/ourscript.js [hunk 9]
```diff
@@ -4,85 +4,215 @@
...
-		var row = $(currentElement).parents('tr');
-		row.children().first().before(`<td>${team}</td>`);
-		row.children().last().remove();
-		row.children().last().after(`<td class="td-actions text-right"><button type="button" rel="tooltip" class="btn btn-success" onclick="deleteRow(this)"><i class="material-icons">how_to_reg</i></button></td>`);	
-		inSessionTable.append(row);
-		
-		clearInterval(timerIntervals[id]);
-		timers[id] = 0;
-		timerIntervals[id] = setInterval(setTime, 1000, id);
+		console.log(currentElement)
```
Comment (frinzekt) [inline app/app/static/custom/ourscript.js:200]: Remove console log

File: app/app/static/custom/ourscript.js [hunk 10]
```diff
@@ -4,85 +4,215 @@
...
-		row.children().last().remove();
-		row.children().last().after(`<td class="td-actions text-right"><button type="button" rel="tooltip" class="btn btn-success" onclick="deleteRow(this)"><i class="material-icons">how_to_reg</i></button></td>`);	
-		inSessionTable.append(row);
-		
-		clearInterval(timerIntervals[id]);
-		timers[id] = 0;
-		timerIntervals[id] = setInterval(setTime, 1000, id);
+		console.log(currentElement)	
+		status = "In Session"
+		clearInterval(id);
```
Comment (frinzekt) [inline app/app/static/custom/ourscript.js:202]: "id" is not an interval. You have to select an item from the timerInterval

File: app/app/static/custom/ourscript.js [hunk 11]
```diff
@@ -134,3 +264,54 @@ $('#librariansAvailableIncrement').click(function () {
...
+			if (element.status == "In Queue") {
+				// Today - `enterQueueTime` is the relative date in seconds
+				const timeDifferenceMiliseconds = (new Date()).getTime() - (new Date(element.enterQueueTime)).getTime()
+
+				// Convert to seconds (rounded)
+				timers[element.id] = Math.round(timeDifferenceMiliseconds / 1000);
+				console.log(new Date())
+				console.log(element)
+				console.log(new Date(element.enterQueueTime))
+				console.log(timers[element.id])
```
Comment (frinzekt) [inline app/app/static/custom/ourscript.js:305]: Clean up this console log

File: app/app/static/custom/ourscript.js [hunk 12]
```diff
@@ -134,3 +264,54 @@ $('#librariansAvailableIncrement').click(function () {
...
+				timerIntervals[element.id] = setInterval(setTime, 1000, element.id);}
+
+			else {
+				const timeDifferenceMiliseconds = (new Date()).getTime() - (new Date(element.changeSessionTime)).getTime()
```
Comment (frinzekt) [inline app/app/static/custom/ourscript.js:309]: Both if statement branch are exactly the same except for which element is being selected. Consider simplifying this by using a ternary operator on which element to select

Eg.
```javascript
const timeElement = element.status=="In Queue" ? "enterQueueTime" : "changeSessionTime"

//... Insert other code here
timeDifferenceMiliseconds = (new Date()). getTime() (new Date(element[timeElement] ) ). getTime()
//... Insert other code here
```

**Review by frinzekt — CHANGES_REQUESTED** [review]
Looks good. Just need a couple of changes for code readability

File: app/app/static/custom/ourscript.js (hunk 10 above)
Comment (Alexkszzz) [reply app/app/static/custom/ourscript.js]: That clearInterval shouldn't be existed, deleted!

File: app/app/static/custom/ourscript.js (hunk 12 above)
Comment (Alexkszzz) [reply app/app/static/custom/ourscript.js]: Great idea! commited.

File: app/app/static/custom/ourscript.js (hunk 7 above)
Comment (Alexkszzz) [reply app/app/static/custom/ourscript.js]: totally not related, the function parameter should only be 'id'. Fixed!

File: app/app/static/custom/ourscript.js (hunk 6 above)
Comment (Alexkszzz) [reply app/app/static/custom/ourscript.js]: I combined them into one using ternary operator. Thanks for the idea xD

File: app/app/static/custom/ourscript.js [hunk 13]
```diff
@@ -4,85 +4,202 @@
...
+			body: JSON.stringify(data),
+			headers: {
+				'Content-Type': 'application/json'
+			},
+		})
+		const dataResponse = await response.json()
+		if (dataResponse.status == "Ended"){
+			const index = queueList[dataResponse.queue].findIndex((element) => element.id == dataResponse.id)
+			// pop the data from current queue
+			queueList[dataResponse.queue].splice(index, 1)
```
Comment (frinzekt) [inline app/app/static/custom/ourscript.js:85]: Do you not need to rerender table after this operation?

**Review by frinzekt — CHANGES_REQUESTED** [review]

**Comment by frinzekt** [comment]
Note @UWA-CITS3200-18-2021/c3200-team , this PR contains a breaking change that will require you to delete all traces of your database to undergo full migration process.

If you have any difficulties after this change, please contact me or @Alexkszzz.

File: app/app/static/custom/ourscript.js (hunk 13 above)
Comment (Alexkszzz) [reply app/app/static/custom/ourscript.js:85]: I put rerender after the if else command
https://github.com/UWA-CITS3200-18-2021/ReSQ/blob/2770643dd6f69f02fe7122243d00a06c474d6e3c/app/app/static/custom/ourscript.js#L87-L94

**Review by frinzekt — APPROVED** [review]
All good!

**Pull request merged**